pkg/template/processor_test.go

```diff
@@ -0,0 +1,807 @@
+			// given
+			values := map[string]string{
+				"USERNAME": user,
+				"COMMIT":   commit,
+			}
+			tmpl, err := decodeTemplate(decoder,
+				createTemplate(withObjects(namespace, rolebinding), withParams(usernameParam, commitParam)))
+			require.NoError(t, err)
+
+			// when
+			objs, err := p.Process(tmpl, values, template.RetainAllButNamespaces)
+
+			// then
+			require.NoError(t, err)
+			require.Len(t, objs, 1)
+			// assert namespace
+			assertObject(t, expectedObj{
+				template: rolebindingObj,
+				username: user,
+				commit:   commit,
+			}, objs[0])
+		})
+
+	})
+}
+
+func TestProcessAndApply(t *testing.T) {
+
+	commit := getNameWithTimestamp("sha")
+	user := getNameWithTimestamp("user")
+
+	s := addToScheme(t)
+	codecFactory := serializer.NewCodecFactory(s)
+	decoder := codecFactory.UniversalDeserializer()
+
+	values := map[string]string{
+		"USERNAME": user,
+		"COMMIT":   commit,
+	}
+
+	t.Run("should create namespace alone", func(t *testing.T) {
+		// given
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s)
+		tmpl, err := decodeTemplate(decoder,
+			createTemplate(withObjects(namespace), withParams(usernameParam, commitParam)))
+		require.NoError(t, err)
+		objs, err := p.Process(tmpl, values)
+		require.NoError(t, err)
+
+		// when
+		err = p.Apply(objs)
+
+		// then
+		require.NoError(t, err)
+		assertNamespaceExists(t, cl, user)
+	})
+
+	t.Run("should create role binding alone", func(t *testing.T) {
+		// given
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s)
+		tmpl, err := decodeTemplate(decoder,
+			createTemplate(withObjects(rolebinding), withParams(usernameParam, commitParam)))
+		require.NoError(t, err)
+		objs, err := p.Process(tmpl, values)
+		require.NoError(t, err)
+
+		// when
+		err = p.Apply(objs)
+
+		// then
+		require.NoError(t, err)
+		assertRoleBindingExists(t, cl, user)
+	})
+
+	t.Run("should create namespace and role binding", func(t *testing.T) {
+		// given
+		cl := test.NewFakeClient(t)
+		p := template.NewProcessor(cl, s)
+		tmpl, err := decodeTemplate(decoder,
+			createTemplate(withObjects(namespace, rolebinding), withParams(usernameParam, commitParam)))
+		require.NoError(t, err)
+		objs, err := p.Process(tmpl, values)
+		require.NoError(t, err)
+
+		// when
+		err = p.Apply(objs)
+
+		// then
+		require.NoError(t, err)
+		assertNamespaceExists(t, cl, user)
+		assertRoleBindingExists(t, cl, user)
+
+	})
+
+	t.Run("should update existing role binding", func(t *testing.T) {
+		// given
+		cl := test.NewFakeClient(t)
+		cl.MockCreate = func(ctx context.Context, obj runtime.Object, opts ...client.CreateOption) error {
+			if rb, ok := obj.(*unstructured.Unstructured); ok {
+				rb.SetResourceVersion("1")
+				return cl.Client.Create(ctx, obj, opts...)
+			}
+			return cl.Client.Create(ctx, obj, opts...)
+		}
+		cl.MockUpdate = func(ctx context.Context, obj runtime.Object, opts ...client.UpdateOption) error {
+			if u, ok := obj.(*unstructured.Unstructured); ok {
```

Comment:
should you cast to `runtime.Unstructured` here?

Reply:
that was already replaced here 657e55c  - you reviewed old code